### ISSUE TITLE
Make it possible to create a receipt for the whitesblower

### ISSUE THREAD

**Issue by pvigoseidor** [globaleaks/GlobaLeaks#3397]
### Proposal

As a whistleblower of GlobaLeaks, I want to be able to download or copy a receipt that includes the following information:
- The entity logo
- A unique ID
- Receipt number
- Date
- Hour
- Information text about the case is opened.

### Motivation and context

The purpose of this new feature is to provide assurance to whistleblowers that their case has been registered. In the event that the entity removes the case or fails to respond, the feature allows the whistleblower to escalate the issue to another entity. This serves as a motivating factor for individuals to come forward and report any wrongdoing without fear of their report being ignored or overlooked.

It will be necessary to discuss the information of the receipt to remain the security of the whistleblower.

**Comment by evilaliv3** [comment]
Thank you @pvigoseidor for this proposal.

I recognize the value of this feature for giving a proof to whistleblower of the fact that they performed a report.
This is a topic well known in the community for which at the moment we have more doubts than answers.

The main two are:
- How should this receipt be implemented to enable a whistleblower to be used as a proof. In other words how can we prevent the receipt to be hand creafted?
- Offering a user such a proof, exposes the whistleblower to the risk the proof is leaked and used against them to disclose the fact that they have made the report. Most of the users will probably for example download the file on their computer or share it in an appropriate manner. For example i point out that the currenct application is specifically designed to avoid when possible to leave forensic evidences on the computer of the whistleblower.

I know that some other users have opinion in favor or in critic of this topic and that some are already working on similar concepts; i tag some her to open the discussion.

@lorenzogentile @giorgiofraschini @elbill

**Comment by evilaliv3** [comment]
Thank you. 

I consider this is definitely a necessity if we implement such a feature and probably we should implement the proposed warning in general for any download mechanism implemented on the whistleblower interface.

In general in relation to the feature my main doubt is about the validity of the receipt and it's sense. Its worthless in my opinion to expose users to risks if there is not an official way to prove the receipt was emitted and assigned specifically to that person for example. How is this receipt expected to be used by the user and when?

**Comment by pvigoseidor** [comment]
A receipt can be used in cases where an entity fails to take action on a newly opened case. It is important to consider cases where the whistleblower wants proof that the case has been opened, especially when the case is not anonymous. 
While it's crucial to protect the anonymity of whistleblowers, providing a receipt can help ensure transparency and accountability in cases where anonymity is not required.

**Comment by pvigoseidor** [comment]
One possible solution to eliminate the need for downloading a file and ensure that the whistleblower is aware of the case being opened is to display a QR code containing the relevant information. The whistleblower can scan the code to access the details and confirm that the case has been opened.

It will be helpful If the whistleblower wishes to retain the information.

**Comment by evilaliv3** [comment]
This article relates both to article 9  and article 21 of the european directive on whistleblowing

The main issue we should probably try to address is: How could we offer the whistleblower a receipt that proves they performed a report? a proper receipt is typically an hash of the report signed by who emits the receipt. here we do not have the possibility nor to sign nor to preserve the original data on which the hash is computed.

\cc @rglauco @larrykind @danielvaknine @gianlucagilardi

**Comment by danielvaknine** [comment]
It is an interesting idea, but as stated I too see a difficulty in making it "verifiable".

As I see it, the basic functionality is fairly close at hand – the **print/export**-functions from the user side could be used as "proof" (even though it would be very difficult to verify/validate).

Then, there could be 2 ways of verifying the report:
1. Admin's audit log where admin can match the report submission date to when a report was created in the platform
2. Creation of a unique Case ID that is saved somewhere for future potential verification (such as admin's audit log or user's interface)

This could be a simpler approach to it. Also, some sort of "proof" (even if it's not impossible to create a fake such) is better than none, in my opinion.

**Comment by larrykind** [comment]
Hello everyone, 
definitely agree with @evilaliv3 concerns about implementing a receipt in a "classic" way, I mean something phisical like a file to be saved in every way, because it could be easily forged.
I think that any way to preserve the report after the deletion, it's clearly not a deletion, so I found a good idea what @giorgiofraschini proposal, to evaluate a "non-deletion" and non changeable grace time to preserve the report: the whistleblower could demonstrate he made the report within that time period using his key-code.

As previoulsy discussed, I think also a good way to proceed could be a deletion-approval workflow that involves all the receivers that can see the report _and the whistleblower also_, in order to speed up the deletion. if the whistleblower will approve the deletion, the report could be immediately wiped. Otherwise, default timers can be used to delete, the upsaid deletion grace time or the normal expiration timers also can be used as a reference.

**Comment by larrykind** [comment]
As an adding to what I previously said, 
I agree that some track about a deleted report should be leaved on the platform, but that track must not contain anyting referring the data inside the report. Sorry to bother.

**Comment by gianlucagilardi** [comment]
I follow @giorgiofraschini and @danielvaknine line of reasoning.

I imagine something like this:
1. upon submission of the report, the whistleblower is requested to provide a strong "password" 
2. the report as such is packed and protected with such password
3. the packed, password protected (encrypted) report is saved remotely (ideally on the GL host) in cold storage and time stamped and cannot be deleted so that recipient(s) cannot tamper with it (thus providing some resilience against tampering - be it for art 17 compliance reasons or malicious tampering)
4. since the packed report is encrypted with a strong password, the data therein could be considered as anonymous from the recipient/platform point of view (this provides GDPR and art. 17 compliance if and when the report has to be redacted to pursue art. 17 compliance)
5. the whistleblower at any time can access the platform like today and should be able to retrieve/download the "original" packed, password-encrypted report saving locally the report itself (with any appropriate warning).

**Comment by evilaliv3** [comment]
Thank you for all your valuable feedback.

I just had a very brief meeting with @rglauco and we have some ideas.
this text states basically what @gianlucagilardi say and is based on some discussion i had with him actually but i took it very long to fomulate it :

------

To offer the possibility to the whistleblower to prove their report the user should be able to preserve online or offline a copy of the report up to their own choice and this is fully in compatibility with GDPR.

Online preservation: Preserving it online offers more guarantees because just require the user to memorize the small receipt of 16 digits that we use for authentication and require the data processor to maintaining encrypted copy of the data without having any access to the key. Wanting to raise the security of this process eventually the system could as well generate a very secure key offer the possibility to download this key or store it as a qrcode during the time of the validity of the 16 digits receipt. Afterwards this longer key would not enable to access the system is data processing. This is the solution i would advice for but it is not fully compatible with GDPR that consider data storage a type of processing even if the processor do not have any copy of the key.

Offline preservation: Mirror implementation of the Online preservation an Offline Preservation could be implemented by offering the Whistleblower the possibility to download an encrypted copy of the original encrypted with a key that is preserved by the data processor. This would be in fully compatibility with GDPR but has a very significant issue that should be wisely considered in the context of whistleblowing: How could we require the whistleblower, in its own poor conditions of whistleblower exposed to stress, download a copy of all the data they just uploaded and provided to them in a non-intelligeble and encrypted form, with a size that is bigger than what they uploaded and ask them to preserve this data somewhere to prove they loaded that data?

**Comment by rglauco** [comment]
Perhaps a summary report could be created with the text of the report and the filenames of any attachments along with the corresponding hashes. Such a report could be encrypted with the whistleblower provided password or signed with a key inaccessible to the receiver and be steganographed into a downloadable image (uploaded by the whistleblower or already in the system). Steganography would allow to upload more data and the data signature can be used to prove the submission. In this way, the whistleblower would have the detail data to substantiate the report, in case of incorrect behavior on the part of the receiver.

The system should be able to update the summary report and to download of the image, in case the whistleblower uploads more documents and textual additions.

**Comment by gianlucagilardi** [comment]
_"that consider data storage a type of processing even if the processor do not have any copy of the key"_

I beg to slightly differ on this.  Trues, it is still processing but it is not necessarlily processing _of personal data_. Following the ECJ Breyer decision encrypted data for which the hosting provider has no reasonable or legal means of decrypting  fall outside the scope of the GDPR  (even if the unencrypted data contain personal data); see also in this respect the data transfer guidelines of the EDPB which listed encryption of data with a key solely in the availability of the data exporter as a sufficient security measure to transfer personal data in jurisdiction where it would otherwise impossible to do so.

**Comment by evilaliv3** [comment]
Thank you @gianlucagilardi; actually this really helps; since our precious discussion i was considering was actual your point to sustain that i still is processing.

This enfocers our position that Online Preservation is best option

Analyzing the Online Preservation more, i realize we will have anyhow a significant issue: if we agree on this way of proceeding we will have store reportings forever, we open the doors of whistleblowing systems to DoS attacks. The administration would have basically to guarantee unlimited data storage and thus providers.

**Comment by gianlucagilardi** [comment]
_"With this in mind during the online preservation stage there is no need for any user password and whistleblower will be able / will need to set that password only when they will decide to download the data."_

@evilaliv3 I might have to disagree with that. Having a plain readable (further) copy in the system defies the compliance purpose, since in this scenario a recipient might need to redact some information which would be anyway accessible in the system (since it is not encrypted because the whistleblower has not requested a copy/download - yet) which would imply a non conformity of art. 17 given - as per above - that lacking the whistleblower-controlled encryption this would definitely be personal data processing.

**Comment by evilaliv3** [comment]
@gianlucagilardi: actually i think there could be some misunderstanding here.

The receipt is never stored in plaintext on the server and in the implementation consider we would never have a plain readable copy of any data on the system.

The data would be always encrypted online using the key derived from the secret receipt in the end of the whistleblower or encrypted offline with the key derived by the secret password set by the whistleblower.

**Comment by gianlucagilardi** [comment]
> @gianlucagilardi: actually i think there could be some misunderstanding here.
> 
> The receipt is never stored in plaintext on the server and in the implementation consider we would never have a plain readable copy of any data on the system.
> 
> The data would be always encrypted online using the key derived from the secret receipt in the end of the whistleblower or encrypted offline with the key derived by the secret password set by the whistleblower.

Then the question is: how easy (i.e., are there "reasonable means") would be for the host manager/recipient to brute force the further copy or reverse the encryption? Those are the Breyer requirements in order to exclude the personal nature of the data at hand.... :)

**Comment by evilaliv3** [comment]
the parameters of argon2 are set to require 1second of computation and occupy 128kb of ram during that second.

10^16 seconds then if you have only 128k of RAM, less if are the NSA

**Comment by evilaliv3** [comment]
I would like to point out that i've improved my current my analysis in comment https://github.com/globaleaks/GlobaLeaks/issues/3397#issuecomment-1517572546 adding as requirement for the offline receipt generation to have a third party to sign the receipt in order to assign it a certain timestamp.

**Comment by evilaliv3** [comment]
I would like to point out that with all the evaluations made today it seems to me that we have addressed al the issues except that we still miss a way to prove the report/receipt where generated in a specific time, that is something that requires an electronic  third party signature verifiable when needed.

**Comment by danielvaknine** [comment]
> It is an interesting idea, but as stated I too see a difficulty in making it "verifiable".
> 
> As I see it, the basic functionality is fairly close at hand – the **print/export**-functions from the user side could be used as "proof" (even though it would be very difficult to verify/validate).
> 
> Then, there could be 2 ways of verifying the report:
> 
> 1. Admin's audit log where admin can match the report submission date to when a report was created in the platform
> 2. Creation of a unique Case ID that is saved somewhere for future potential verification (such as admin's audit log or user's interface)
> 
> This could be a simpler approach to it. Also, some sort of "proof" (even if it's not impossible to create a fake such) is better than none, in my opinion.

These suggestions are in our opinion much simpler and straightforward and could still be verifiable. Some other suggestions could be both difficult to implement and use/explain in a simple way.

**Comment by elbill** [comment]
Joining the conversation a bit late. Following also a recent chat we had the simplest way to achieve this is to provide the whistleblower with a unique string/number as a protocol number. Also this has been set a a very desirable feature of our NAtional Transparency Authority (former anti corruption authority).
There should be a way to connect this number withe the serial number of the report. The automatically generated protocol number could be included in the report as an additional "question" that both recipient and whistleblower can access so the recipient can identify this with the search function. Regarding the proof the whistleblower has access to their report and can anyway take screenshot or print it. O would not object (would actually like) the idea of a print button in the whistleblower interface.
So in short
1) generation of unique protocol number and inclusion in both whistleblower and recipient report view. Searchable for recipient
2) print button for the whistleblower. they will be able to print or generate pdf of the timestamped report and the serial number.
I believe this is the simplest way to implement this.

This does not serve as a systemic proof of recept. For this we should keep in audit log a copy of the protocol number of the deleted reports to validate that a report has been received.
I think this feature in its simplified form will not take so many resources. We can assist with this.
Hope this helps

**Comment by danielvaknine** [comment]
Finding your ideas interesting but would like to highlight that we also
need to take in mind user friendliness for the average user/recipient, i.e.
not add too many functions etc to the recipient interface.

Perhaps exploring your bullet point 1. but with only accessing this from
the admins audit log.

Just a thought.

**Comment by evilaliv3** [comment]
Thank you.

I understand the reasons and generating a uuid4 to be shown to both
whistleblowers and recipients that identify the reports makes sense.

Here some aspects to consider:
- a random uuid4 exposed is valid to support this use case. Anyhow if it is
exposed to all recipients is not a proof of anything. Recipients and Admins
having access to it can engage anytime with a fake whistleblower that will
testimony to have performed a fake report while the real report had been
deleted. It should probably be encrypted and shown only to recipientsm we
- the whistleblower could possibly be confused having two numbers (this ID
and the receipt)
- printing the 16 digits receipts has been till now discouraged to avoid
risks of leaking the receipt to others. Suggesting it is a big change that
could result significantly misleading

**Comment by elbill** [comment]
@evilaliv3 the whistleblower will print their report after accessing it.
So will not print the 16 digit number.
The print button (and the uuid4 will be placed there and teh whistleblower will have the right to print their report)
Even if we want to place it in the confirmation screen we can hide the 16digit number.
I see your point with deleted reports.
This solution is if in the audit log the deleted report entry is not deleted but remains as status "Deleted". This is related to our discussion but should be implemented anyway. When deleting the report , deleting its audit log entry is like interfering witht the audit log. Any thrace of the report (when happened) is gone. This info is also not high risk and could remain in hte system even when report content is deleted.

1) generation of unique protocol number and inclusion in both whistleblower and recipient report view (not confirmation screen). Searchable for recipient. Teh whistleblower will not need to remember any second number. Whenever they want they can print their report - including the uuid4
2) In the audit log we do not delete the entry of a deleted report but put a flag or status as deleted. This is a good practice anyway. When a user deletes a report, it is logged in their log but we can not know which report that was.
3) print button for the whistleblower. They can print their page anyway, make it easyer for them. 

@danielvaknine I do not think that any of these steps is additional work for the WB or the recipient. It just gives the possibility for confirmation if and when they need it. 
In a case that I made a report, and do not hear from the company or authority, how can I prove that I made the report. And if the report is deleted against my will what can I do to prove it?

**Comment by danielvaknine** [comment]
Really agree that every trace of a report shouldn't be deleted in the audit
log, but rather be marked as "Deleted".

**Comment by elbill** [comment]
**Coming back to the issue because I believe it is really important.**
Reviewing the discussion I suggest to implement:

1. A print/export button in the whistleblower report interface, enabled only after acknowledgment of a risk disclaimer with the use HTML + JavaScript snippet to implement a “Print as PDF” button that saves the report content as a PDF using the browser’s built-in print-to-PDF functionality. 
2. The exported document (PDF or print view) to exclude the 16-digit access code and instead include:

- Report UUID or protocol number (non-sensitive, verifiable if audit log entry exists)
- Timestamp of submission
- Optional watermark: “Generated by GlobaLeaks on [date/time]”
-  Hash summary of report content and filenames, not the content itself, to preserve confidentiality but still allow integrity verification if needed.

This would align with proposals already discussed (e.g. UUID display, confirmation of submission, audit log traces) while:
	•	Avoiding offline storage of sensitive data
	•	Respecting GDPR/Article 17 concerns
	•	Enabling user-friendly, low-barrier documentation for whistleblowers

**This approach can coexist with future implementations of verifiable receipts or online preservation mechanisms.**

**Comment by JaviAlama** [comment]
Good morning, has there been any progress on this matter? Our organization has identified a need for whistleblowers to be able to obtain a PDF copy of their report, at the moment of sending it.
Just as there is a “View your report” button, could a “Download your report as a PDF” button be added there?

Thank you

**Comment by evilaliv3** [comment]
Hello @JaviAlama 

thank you for your feedback and suggestion.

There has been no progress, but i think your suggestion is a very valid one.

The implementation is not as simple as you might think but we could try to make an attempt to implement it later in the year.

We should:
- make it possible to download the report in the whistleblower language; as you know in fact the current PDF included inside the export is not fully translated.
- as for the text i think we can just use "download" that is already translated in every language


For the moment the best recommendation is to suggest users to use the "Print" functionality that already enable to print the report in the language currently selected by the user.
